### PR TITLE
feat(send_queue): Persist failed to send errors

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Modify the SendQueue in order to persist the error that cause the event to fail to send as a `QueueWedgeError`. The `QueueWedgeError` is not a 1:1 mapping for all kinds of errors, but holds variant and information that the client can react to in order to propose "quick fixes"/solution before retrying to send.

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/3973 
Also fixes https://github.com/element-hq/element-x-ios/issues/3287 because when a timeline reset occurs the fail to send reason is also lost.

This PR starts with a refactoring commit https://github.com/matrix-org/matrix-rust-sdk/commit/e7696003e846b64c41761109f326fd37c4506040 to introduce the new `QueueWedgedError` and move the logic that was in the ffi layer to convert api errors to SendState error variant. This `QueueWedgedError` can be directly use in the `SendingFailed` variant and expose to ffi.

Second commit https://github.com/matrix-org/matrix-rust-sdk/commit/109c1337465ba7965825fec858fd2a4b8b954611 adds the persistence, `QueuedEvent` now have an optional error field instead of a `is_weged` boolean. Same for LocalEchoContent::Event.
Adds also Migration for sqlite and indexeddb

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-base/src/store/traits.rs
Comment (jmartinesp) [inline crates/matrix-sdk-base/src/store/traits.rs:1160]: Should we update this doc comment, since an identical one was added below to `is_wedged()`?

File: crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql
Comment (jmartinesp) [inline crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql:4]: I don't know how this works in Rust, would this still work if `QueueWedgeError` adds/removes some case?

**Review by jmartinesp — COMMENTED** [review]
LGTM in general, but some clarification about the serialization/deserialization of the BLOB in the DB when there are new/removed enum cases would be good.

Also, since this is related to the send queue, maybe a 2nd pair of eyes from @bnjbvr wouldn't hurt either.

File: crates/matrix-sdk-ui/src/timeline/controller/mod.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/controller/mod.rs]: +1, let's inline this.

File: crates/matrix-sdk-ui/src/timeline/controller/mod.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/controller/mod.rs]: nit: for `&str` to `String` conversions, we prefer `.to_owned()`.

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:719]: Outside the FFI layer, we're not using stringly-typing, so this would become a `HashMap<OwnedUserId, Vec<OwnedDeviceId>>`, and similarly below.

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:714]: Since this is related to the send queue, can we put it in the send queue file? Or is there a reason I'm missing to put it into deserialized_responses?

Also, should this use our common pattern of deriving `thiserror::Error`, to make it clear it's an error type, and would avoid the manual display impl? (uniffi also has an `uniffi::Error` derive which could be convenient here)

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:725]: `Vec<OwnedUserId>`

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:731]: Could it be a `Box<dyn Error>` or something similar? It's a bit weird to store the stringified error, since API users can't match on it, unless they precisely know what string to expect.

File: crates/matrix-sdk/tests/integration/send_queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/send_queue.rs:475]: This change justifies why we need to store the error itself, in the case of `GenericApiError`: otherwise here we're loosing information that it could be a client API error. Maybe `GenericApiError` could contain a (boxed or not) `matrix_sdk::Error`, as this is the most general error type we have in this crate.

File: crates/matrix-sdk-base/src/store/traits.rs
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/store/traits.rs:396]: Can you clarify the contract around what it means when `error` is not set? (I suppose that means "unwedge", but it's not specified here, and not obvious at first sight)

File: crates/matrix-sdk-base/src/store/traits.rs
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/store/traits.rs:1164]: nit: this is context around what it means to have an error; I'd rather put that near the `error` field itself, and have the doc comment here say something like "returns whether the queued event is wedged (i.e. ran into an unrecoverable error) or not"

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs
Comment (bnjbvr) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:1454]: This doesn't correctly handle the case where legacy is set to false, but there's still a new-form error. I think you want something like:

```rust
Some(true) => // migrate,
_ => item.error
```

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs
Comment (bnjbvr) [inline crates/matrix-sdk-indexeddb/src/state_store/mod.rs:439]: Does the conversion from bool to `Option<bool>` work? (i.e. is there a test for that?)

I think it may work because we're using JSON as the serialization format, but it's not guaranteed it would work in general with any serialization format, so I'd rather have a test for this please.

File: crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql
Comment (bnjbvr) [reply crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql:4]: Yeah, the value is serialized (I suspect to JSON bytes) before being put into the table, and it has to be a blob because of the encryption case.

File: crates/matrix-sdk-sqlite/src/state_store.rs
Comment (bnjbvr) [inline crates/matrix-sdk-sqlite/src/state_store.rs:293]: Nice! Good job on the migration.

File: crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql
Comment (bnjbvr) [reply crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql:4]: Is there no need to annotate this to mark it nullable? and that the default value is null when not set? Maybe it's the default with sqlite, and I don't know, but might be worth indicating in the comment then?

File: crates/matrix-sdk-sqlite/src/state_store.rs
Comment (bnjbvr) [inline crates/matrix-sdk-sqlite/src/state_store.rs:1788]: uber-nit: since the transpose is not trivial, can you add a newline before and after, so we notice there's something going on here, please? (and it'll make the code breathe :))

File: crates/matrix-sdk-sqlite/src/state_store.rs
Comment (bnjbvr) [inline crates/matrix-sdk-sqlite/src/state_store.rs:2273]: I think we can get rid of the local binding and clone here.

File: crates/matrix-sdk/src/send_queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk/src/send_queue.rs:607]: nit: unnecessary qualification here

File: crates/matrix-sdk/src/send_queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk/src/send_queue.rs:1178]: Let's keep `unrecoverable` here. It has a precise meaning in the context of this file, i.e. retrying would lead to the same error again and again (vs network errors that may be transient).

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Good job overall! I have issues with the first commit which loses us a bit of static typing and checkability (because of the use of stringly typed values and errors), so I'd like to take another look, otherwise the code looks great and most of my comments should be minor. Thanks!

**Comment by BillCarsonFr** [comment]
> Good job overall! I have issues with the first commit which loses us a bit of static typing and checkability (because of the use of stringly typed values and errors), so I'd like to take another look, otherwise the code looks great and most of my comments should be minor. Thanks!

Yeah, I started by doing this refactoring because persisting all the different kind of [errors](https://github.com/matrix-org/matrix-rust-sdk/blob/a901506a536bf12b081eea6d6c70777d0fdcc565/crates/matrix-sdk/src/error.rs#L257) required to make a lot of things serializable, and at the end this is not used at all by the application layer (that was doing a mapping of the error to usable SendingFailed types).
What the app was using was more the thing that we really interesting into persisting. 
It's not really the error that the app layer really need, I think of it as a suggestion a compiler could give you when compilation fails. Lik "Re-verify this user or whitelist the problematic device" then retry.

I could have done without the `stringly typed values` and just put a `GenericUnrecoverableError`, the only suggestion is to delete the message. The reason why doesn't matter much as it cannot be presented to the user, it will be an english only technical string.

**Comment by bnjbvr** [comment]
Good point about needing to ser/de all the error types. In that case, can we keep the String for the generic unrecoverable error case, and use proper types for the other error variants?

**Comment by BillCarsonFr** [comment]
> Good point about needing to ser/de all the error types. In that case, can we keep the String for the generic unrecoverable error case, and use proper types for the other error variants?

Hello @bnjbvr, What do you think about the following changes:
So we are looking at all the errors that could happen, and we are doing 2 things:
 - Signalling the errors via `RoomSendQueueUpdate` and `SendQueueRoomError` and telling if is recoverable or not.
 - Persisting the errors that are unrecoverable in the queue.
 
 So my idea is that we keep the global sdk `Error` type for signalling. But for unrecoverable errors that we want to persist we use the new `QueueWedgeError` (that will stringify some types not really useful).
 
 See this commit for this change: https://github.com/matrix-org/matrix-rust-sdk/pull/4137/commits/e824fd24d273319ecb15116bec7384b6b8dbc842
 
 As requested I also put back the ruma OwnedUserId/DeviveId inside the QueueWedgeError variant instead of raw string. I did that initially so that I can directly create ffi bindings. But I see the point, and now I separated 2 versions of the error with a string one for bindings.
 
 https://github.com/matrix-org/matrix-rust-sdk/pull/4137/commits/3e657538a61fd298255504f75b9b2c20fdd75b82

File: crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql
Comment (BillCarsonFr) [reply crates/matrix-sdk-sqlite/migrations/state_store/007_a_send_queue_wedge_reason.sql:5]: Added some comments. And yes the default in NULLABLE and NULL.

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-common/src/deserialized_responses.rs]: see https://github.com/matrix-org/matrix-rust-sdk/pull/4137/commits/3e657538a61fd298255504f75b9b2c20fdd75b82

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-common/src/deserialized_responses.rs:712]: Send Queue is in `matrix-sdk` and this error is persisted in the state storage `matrix-sdk-base`. 
So it cannot be in the send_queue.rs file.

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-common/src/deserialized_responses.rs]: see here https://github.com/matrix-org/matrix-rust-sdk/pull/4137#issuecomment-2425914825

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-common/src/deserialized_responses.rs:733]: see here https://github.com/matrix-org/matrix-rust-sdk/pull/4137#issuecomment-2425914825

File: crates/matrix-sdk/tests/integration/send_queue.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/send_queue.rs]: Fixed now, see here details https://github.com/matrix-org/matrix-rust-sdk/pull/4137#issuecomment-2425914825

File: crates/matrix-sdk/src/send_queue.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk/src/send_queue.rs]: Outdated now, but it was previously because of the `#[cfg(feature = "e2e-encryption")]`. Other option could be to add the feature requirement at the import level (if not would be unused import)

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:715]: With this error message, you won't see the `user_device_map` in the display impl. Saying here in case it's not intended; to display them, you'd need to add something like `: {user_device_map:?}` to the string above, and it will include the debug display of the btreemap.

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:721]:
```suggestion
    /// the current encryption setting prohibits sharing when it happens.
```

File: crates/matrix-sdk-common/src/deserialized_responses.rs
Comment (bnjbvr) [inline crates/matrix-sdk-common/src/deserialized_responses.rs:727]:
```suggestion
    /// It is required to set up cross-signing and properly verify the current
```

File: crates/matrix-sdk/src/send_queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk/src/send_queue.rs:592]: nit: using a `From` impl makes it part of the public API of `QueueWedgeError`, which we probably don't want to do, since it's an internal conversion. Can you use a plain function instead, please?

Also, take the `crate::Error`  by ownership (i.e. no ref `&`), since this could avoid some cloning if I'm reading the code correctly.

File: crates/matrix-sdk/src/send_queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk/src/send_queue.rs:1174]: I might have misunderstood your intent, I thought the goal was to use a `matrix_sdk::Error` here, since it's the most generic type for signalling the error to the caller. Can we do that?

File: crates/matrix-sdk-ui/tests/integration/timeline/queue.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/queue.rs:287]: (with respect to the previous comment, this would mean undoing this, since we could still pattern-match the right client API error here)

File: crates/matrix-sdk-ui/src/timeline/tests/echo.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/tests/echo.rs:68]: uber-nit: let's rename `some_io_error` to `some_error` :nerd_face:

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks for the changes, this is on the right track. I might have misunderstood the intent to store the most generic error type when signalling the error to callers, and it seems we can achieve this; can you try it out please? :)

File: crates/matrix-sdk-indexeddb/src/state_store/mod.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-indexeddb/src/state_store/mod.rs:439]: added here https://github.com/matrix-org/matrix-rust-sdk/pull/4137/commits/93c452872cae5915fa97b7f8eed9c55150d7adcf

File: crates/matrix-sdk/src/send_queue.rs
Comment (bnjbvr) [reply crates/matrix-sdk/src/send_queue.rs:1174]: Why not? After deserializing a `QueueWedgeError`, isn't there a way to turn it into a `matrix_sdk::Error`, by the way of conversion?

Comment (bnjbvr) [reply crates/matrix-sdk/src/send_queue.rs:1171]: Answering myself: no, and prior to the patch we only had the bool "is wedged or not", so this is already quite an improvement. And indeed in other contexts that we care more about, we can keep the `crate::Error`, so the current code is fine as is. Resolving this comment, thanks for the ping :)

File: crates/matrix-sdk/src/send_queue.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk/src/send_queue.rs:589]: It is used in the timeline controller to update the in memory local event EventSendState. Why is it a issue that it is a public API?

File: crates/matrix-sdk-ui/tests/integration/timeline/queue.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-ui/tests/integration/timeline/queue.rs:287]: Not for this one see https://github.com/matrix-org/matrix-rust-sdk/pull/4137#discussion_r1808911747
EventSendState is using `QueueWedgeError` and not the full `crate::Error`. Notice that for this particular error we propbably will want to add a proper variant for `QueueWedgeError` to have a proper media local echo feedback

**Comment by bnjbvr** [comment]
So as to help landing this PR, I've pushed a small commit that keeps the whole `matrix_sdk::Error` for the timeline, so that we don't lose any information that we shouldn't have to lose. At the FFI layer it's always converted into a `QueueWedgeError`, which makes it easy to pattern-match on the other side of the fence, so it's functionally equivalent. With this patch, I don't have more objections to landing this, but before giving my final approval, I'd like to make sure there's no changes required on our side to address the complement crypto bindings failure.

**Review by bnjbvr — APPROVED** [review]
(Approving assuming no dramatic changes for complement crypto)

Thanks for the PR, notably good job getting through all the layers!

**Pull request merged**
